File: 345.reverse-vowels-of-a-string.java
/*
 * @lc app=leetcode id=345 lang=java
 *
 * [345] Reverse Vowels of a String
 */

// @lc code=start
class Solution {
    public String reverseVowels(String s) {
        char[] c = s.toCharArray();
        int l = c.length;
        int left = 0;
        int right = l - 1;
        HashSet<Character> vowel = new HashSet<>();
        for (char a : "aeiouAEIOU".toCharArray()) {
            vowel.add(a);
        }
        while (left <= right) {
            if (!vowel.contains(c[left])) {
                left++;
            }
            else if (!vowel.contains(c[right])) {
                right--;
            }
            else {
                char temp = c[left];
                c[left] = c[right];
                c[right] = temp;
                left++;
                right--;
            }
        }
        return new String(c);
    }
}
// @lc code=end

File: 345.reverse-vowels-of-a-string.py
#
# @lc app=leetcode id=345 lang=python
#
# [345] Reverse Vowels of a String
#

# @lc code=start
class Solution(object):
    def reverseVowels(self, s):
        """
        :type s: str
        :rtype: str
        """
        vowel = set('aeiouAEIOU')
        n = len(s)
        a = list(s)
        left = 0
        right = n - 1
        while left <= right:
            if a[left] not in vowel:
                left += 1
            elif a[right] not in vowel:
                right -= 1
            else:
                a[left], a[right] = a[right], a[left]
                left += 1
                right -= 1
        return ''.join(a)
        
# @lc code=end

